sdl_viewer/src/graphic/moving_window_texture.rs

```diff
@@ -0,0 +1,364 @@
+    /// handled to keep the code from becoming more complex. Also, either may
+    /// be empty, which is the case for purely horizontal/vertical shifts.
+    ///
+    /// These strips have to be inserted correctly into the OpenGL texture,
+    /// which starts not at 0 but at `texture_offset` because of wrapping indexing.
+    /// Updating that texture from such a strip can typically not be done in one
+    /// step, but requires breaking it into up to four subregions. This is
+    /// because OpenGL textures can be read with wrapping indexing, but not
+    /// written with wrapping indexing, so we need to do it ourselves.
+    /// Let's look at the vertical strip:
+    ///
+    /// ```text
+    /// <- width ->
+    /// +---+-+
+    /// | 3 |4|
+    /// +---+-+
+    /// |   | |  vert_strip
+    /// | 1 |2|
+    /// |   | |
+    /// +---+-+
+    /// ```
+    ///
+    /// Wraparound at the right texture edge causes the split into regions 1/3
+    /// vs 2/4, and wraparound at the upper texture edge causes the split into
+    /// regions 1/2 vs 3/4.
+    /// Here's the texture with the vert_strip placed correctly:
+    ///
+    /// ```text
+    /// <- self.size ->
+    /// +-+-------+---+
+    /// | |       |   |
+    /// |2|       | 1 |
+    /// | |       |   |
+    /// +-+       +---+ <- texture_offset.y
+    /// |4|       | 3 |
+    /// +-+-------+---+
+    ///           ^
+    ///           |
+    ///           xoff
+    /// ```
+    ///
+    /// `xoff` is not equal to `texture_offset.x` if the texture moved right.
+    ///
+    /// The way this is handled is by creating regions as if they always need
+    /// splitting, and skipping empty regions that result from that.
+    pub fn incremental_update(
+        &mut self,
+        delta_x: i32,
+        delta_y: i32,
+        vert_strip: ImageBuffer<P, Vec<P::Subpixel>>,
+        hori_strip: ImageBuffer<P, Vec<P::Subpixel>>,
+    ) {
+        // width/height are always positive, but if the delta is negative, the
+        // start of the new region is actually offset + delta, not offset
+        let xoff = if delta_x > 0 {
+            self.u_texture_offset.value.x
+        } else {
+            (self.u_texture_offset.value.x + delta_x).mod_floor(&self.size)
+        };
+        let yoff = if delta_y > 0 {
+            self.u_texture_offset.value.y
+        } else {
+            (self.u_texture_offset.value.y + delta_y).mod_floor(&self.size)
+        };
+
+        // u_texture_offset.value is always in [0; self.size)
+        let x = (self.u_texture_offset.value.x + delta_x).mod_floor(&self.size);
+        let y = (self.u_texture_offset.value.y + delta_y).mod_floor(&self.size);
+        self.u_texture_offset.value = Vector2::new(x, y);
+
+        let vert_regions = UpdateRegion::new_regions(xoff, y, self.size, &vert_strip);
+        let hori_regions = UpdateRegion::new_regions(x, yoff, self.size, &hori_strip);
+        let regions = ArrayVec::from(vert_regions)
```

Comment:
Do we need `ArrayVec` here? Wouldn't `let regions = vert_regions.iter().chain(&hori_regions)` work?

Reply:
You're right, in a previous iteration of the loop the regions owned the pixels so I had to use `ArrayVec` but now they don't anymore.